### DISCUSSION TITLE
Preprocessing Dataset

### DISCUSSION THREAD

**Discussion by Safqut** [MDIL-SNU/SevenNet#113] [Q&A]
I ran a simulation in lammps to generate a .extxyz file that I wanted to use for training the model. These were my commands in lammps input script to get the **.extxyz** file:
"reset_timestep 0
timestep ${tmstep}
thermo_style custom step time temp pe ke etotal press vol

compute         1 w msd com yes
compute         2 h msd com yes
compute         3 he msd com yes
compute         myPE all pe/atom
compute         myStress all stress/atom NULL

fix    1 all npt temp ${myT} ${myT} ${tdamp} aniso ${myP} ${myP} ${pdamp}
fix    3 all ave/time 1 10 100 v_T v_E c_peW c_peH c_peHe v_V v_Lx v_Ly v_Lz v_P c_p[1] c_p[2] c_p[3] c_1[4] c_2[4] c_3[4] v_W2 v_H v_He
fix    4 all ave/time 1 400 4000 v_T v_E c_peW c_peH c_peHe v_V v_Lx v_Ly v_Lz v_P c_p[1] c_p[2] c_p[3] c_1[4] c_2[4] c_3[4] v_W2 v_H

dump                  1 all custom 1 safqut.extxyz id type x y z fx fy fz c_myPE c_myStress[1] c_myStress[2] c_myStress[3] c_myStress[4] c_myStress[5] c_myStress[6]
dump_modify     1 format line "%d %d %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f"
dump_modify     1 element W H He  # Specify elements if needed

run 1000"

After getting this file, I ran **"sevenn_graph_build safqut.extxyz 5.0"** to build the proper **"graph.pt"** file for training. But in the **"graph.yaml"** file I see this output:
"cutoff: 5.0
files:
- /scratch/ssanwar/LAMMPS2/safqut.extxyz
num_atoms:
  H: 1986985
  He: 10010
  Li: 5005
  total: 2002000
num_graphs: 1001
sevennet_version: 0.10.0
species:
- H
- He
- Li
statistics:
  force_of_atoms:
    max: 54.275535583496094
    mean: 2.5405910014653088e-11
    median: -0.0009694999898783863
    min: -53.439876556396484
    std: 0.7563928365707397
  num_neighbor:
    max: 40.0
    mean: 26.156173706054688
    median: 26.0
    min: 22.0
    std: 0.8698360323905945
  per_atom_energy:
    max: 0.0
    mean: 0.0
    median: 0.0
    min: 0.0
    std: 0.0
  stress:
    max: .nan
    mean: .nan
    median: .nan
    min: .nan
    std: .nan
  total_energy:
    max: 0.0
    mean: 0.0
    median: 0.0
    min: 0.0
    std: 0.0
when: 2024-10-28 16:34
"

My issues are, firstly, the command is not detecting the **correct types of atoms**, it is assuming **H, He, and Li** whereas my actual atoms are **W, H, and He**. Secondly, it is not reading the **energy** or the **stress** from the **.extxyz** file. The energy is showing **0** and stress is showing **nan**. Please help me get this data processing part right! Thank you.

**Comment by YutackPark** [comment]
Thanks for reaching me out. Correctly reading lammps dump file using ASE is sometimes tricky. I need to know how actually reads your data.

```python
import ase.io
atoms = ase.io.read("path to your extxyz", index=':')
print(atoms.info)
print(atoms.arrays.keys())
print(atoms.calc.results)
```

Could you share above results?

**Reply by Safqut** [reply]
Hello. Thank you for replying. I ran your script both interactively and by saving the script. In either case I get this error:
Traceback (most recent call last):
  File "/scratch/ssanwar/LAMMPS2/checklammpsdump.py", line 3, in <module>
    print(atoms.info)
AttributeError: 'list' object has no attribute 'info'

(testenv) [ssanwar@node0376 LAMMPS2]$ python checklammpsdump.py
Traceback (most recent call last):
  File "/scratch/ssanwar/LAMMPS2/checklammpsdump.py", line 4, in <module>
    print(atoms.arrays.keys())
AttributeError: 'list' object has no attribute 'arrays'

(testenv) [ssanwar@node0376 LAMMPS2]$ python
Python 3.10.15 (main, Oct  3 2024, 07:27:34) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ase.io
>>> atoms = ase.io.read("safqut.extxyz", index=':')
>>> print(atoms.info)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute 'info'
>>> print(atoms.arrays.keys())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute 'arrays'
>>> print(atoms.calc.results)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute 'calc'

**Reply by YutackPark** [reply]
Sorry, my fault. you need to access atoms object as `atoms[0]`, not `atoms`.

Like,
```python
print(atoms[0].info)
print(atoms[0].arrays.keys())
print(atoms[0].calc.results)
```

**Reply by Safqut** [reply]
I got this:
(testenv) [ssanwar@node0381 LAMMPS2]$ python checklammpsdump.py
{}
dict_keys(['numbers', 'positions', 'c_myPE', 'c_myStress[1]', 'c_myStress[2]', 'c_myStress[3]', 'c_myStress[4]', 'c_myStress[5]', 'c_myStress[6]'])
{'energy': 0.0, 'forces': array([[ 0.928196,  0.721244,  0.85373 ],
       [-0.067321, -1.506675,  0.817504],
       [-0.659009, -0.078088, -0.204694],
       ...,
       [-0.821   , -0.893346, -0.469909],
       [-0.257858,  0.870212,  1.751953],
       [ 0.254507, -3.231278,  0.779208]])}

**Reply by YutackPark** [reply]
Now I see what happens.

1. For energy, 
```
sevenn_graph_build my_train_data.extxyz 5.0 --kwargs energy_key='c_myPE'
```
This should work.

2. For stress, unfortunately you have to preprocess them in python. It should be inside the `info` dictionary for each atoms object, in eV/Angstrom^3 unit, and list of length 6 following Voigt notation.

3. The species is a very common error when reading LAMMPS. For example, if it is a LAMMPS data (input structure), there is a `Z_of_type` argument: https://wiki.fysik.dtu.dk/ase/ase/io/formatoptions.html#lammps-data
If it is not specified, ASE reads element stating from H, like your results.

But this is for LAMMPS data case, not LAMMPS dump text file (they are not the same). Try `dump_modify` with `element` keyword in LAMMPS (https://docs.lammps.org/dump_modify.html)

Of course, if you are willing to do preprocess atoms object, you can directly modify it in python to have correct element.

+ in sevenn_graph_build, keyword arguments are directly passed to `ase.io.read`. It is a simple surrogate for io.read function, followed by graph build routine.

**Reply by Safqut** [reply]
Thank you. I am going to try these solutions and see if they work. Also If I don't want to use the stress, will it be a problem?

**Reply by Safqut** [reply]
Hello, 
I modified my dump file like this:
"compute         energy all pe/atom
compute         myStress all stress/atom NULL

fix             1 all npt temp ${myT} ${myT} ${tdamp} aniso ${myP} ${myP} ${pdamp}
fix             3 all ave/time 1 10 100 v_T v_E c_peW c_peH c_peHe v_V v_Lx v_Ly v_Lz v_P c_p[1] c_p[2] c_p[3] c_1[4] c_2[4] c_3[4] v_W2 v_H v_He
fix             4 all ave/time 1 400 4000 v_T v_E c_peW c_peH c_peHe v_V v_Lx v_Ly v_Lz v_P c_p[1] c_p[2] c_p[3] c_1[4] c_2[4] c_3[4] v_W2 v_H

dump            1 all custom 1 safqut.extxyz id type x y z fx fy fz c_energy
dump_modify     1 format line "%d %d %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f"
dump_modify     1 element W H He  # Specify elements if needed"

then I ran this command "**sevenn_graph_build safqut.extxyz 5.0 --kwargs energy_key='c_energy'**"

But I got this error:
"Traceback (most recent call last):
  File "/home/ssanwar/.local/bin/sevenn_graph_build", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/ssanwar/.local/lib/python3.11/site-packages/sevenn/main/sevenn_graph_build.py", line 62, in main
    graph_build.build_sevennet_graph_dataset(
  File "/home/ssanwar/.local/lib/python3.11/site-packages/sevenn/scripts/graph_build.py", line 25, in build_sevennet_graph_dataset
    db = SevenNetGraphDataset(
         ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssanwar/.local/lib/python3.11/site-packages/sevenn/train/graph_dataset.py", line 111, in __init__
    super().__init__(
  File "/home/ssanwar/.local/lib/python3.11/site-packages/torch_geometric/data/in_memory_dataset.py", line 81, in __init__
    super().__init__(root, transform, pre_transform, pre_filter, log,
  File "/home/ssanwar/.local/lib/python3.11/site-packages/torch_geometric/data/dataset.py", line 115, in __init__
    self._process()
  File "/home/ssanwar/.local/lib/python3.11/site-packages/torch_geometric/data/dataset.py", line 260, in _process
    self.process()
  File "/home/ssanwar/.local/lib/python3.11/site-packages/sevenn/train/graph_dataset.py", line 141, in process
    tmplist = SevenNetGraphDataset._file_to_graph_list(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssanwar/.local/lib/python3.11/site-packages/sevenn/train/graph_dataset.py", line 391, in _file_to_graph_list
    graph_list = SevenNetGraphDataset._read_ase_readable(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssanwar/.local/lib/python3.11/site-packages/sevenn/train/graph_dataset.py", line 338, in _read_ase_readable
    atoms_list = dataload.ase_reader(filename, **ase_kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssanwar/.local/lib/python3.11/site-packages/sevenn/train/dataload.py", line 320, in ase_reader
    return _set_atoms_y(atoms_list, energy_key, force_key, stress_key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssanwar/.local/lib/python3.11/site-packages/sevenn/train/dataload.py", line 272, in _set_atoms_y
    atoms.info['y_energy'] = atoms.info[energy_key]
                             ~~~~~~~~~~^^^^^^^^^^^^
KeyError: 'c_energy'
"

Then I ran the command "**sevenn_graph_build my_train_data.extxyz 5.0**" without the "**--kwargs energy_key='c_energy'**" part to check if the atoms/elements are atleast correctly read. But unfortunately the "**graph.yaml**" still showing that the atoms are H, He, and Li instead of shwoing that they are W, H, and He. So the "**dump_modify element**" did not work I suppose.

**Reply by YutackPark** [reply]
> I got this: (testenv) [ssanwar@node0381 LAMMPS2]$ python checklammpsdump.py {} dict_keys(['numbers', 'positions', 'c_myPE', 'c_myStress[1]', 'c_myStress[2]', 'c_myStress[3]', 'c_myStress[4]', 'c_myStress[5]', 'c_myStress[6]']) {'energy': 0.0, 'forces': array([[ 0.928196, 0.721244, 0.85373 ], [-0.067321, -1.506675, 0.817504], [-0.659009, -0.078088, -0.204694], ..., [-0.821 , -0.893346, -0.469909], [-0.257858, 0.870212, 1.751953], [ 0.254507, -3.231278, 0.779208]])}

I got it. I misunderstood your outputs. The `info` dict of your `atoms` were empty, instead `arrays` are full of results.

Firstly, if you correctly applied dump_modify, it should work. Here's the example I reproduced.

```bash
(svn_dev) ~/K/p/b/K/S/bulk_md_lmp $ head dump.traj -n 20
ITEM: TIMESTEP
0
ITEM: NUMBER OF ATOMS
72
ITEM: BOX BOUNDS xy xz yz pp pp pp
-4.9583179399999997e+00 9.9166358900000002e+00 -4.9583179399999997e+00
0.0000000000000000e+00 8.5880586000000001e+00 0.0000000000000000e+00
0.0000000000000000e+00 1.0912215430000000e+01 0.0000000000000000e+00
ITEM: ATOMS id element type x y z c_peratom fx fy fz
1 Si 1 2.32913 0 3.63741 -8.13277 0.0244503 1.00874e-06 -3.37372e-06
2 Si 1 2.32913 0 9.09351 -8.13277 0.0244481 3.27744e-06 -2.34088e-06
3 Si 1 -0.150027 4.29403 3.63741 -8.13277 0.0244513 -1.59338e-06 -9.02452e-07
4 Si 1 -0.150027 4.29403 9.09351 -8.13277 0.024452 1.46613e-06 2.29967e-06
5 Si 1 7.28745 0 3.63741 -8.13277 0.0244474 -7.22759e-06 -6.88387e-06
6 Si 1 7.28745 0 9.09351 -8.13277 0.0244474 -4.76057e-06 -5.33182e-06
7 Si 1 4.80829 4.29403 3.63741 -8.13277 0.0244509 -1.22795e-06 -2.27522e-06
8 Si 1 4.80829 4.29403 9.09351 -8.13277 0.0244516 1.62376e-06 1.01374e-06
9 Si 1 -1.16457 2.01709 1.8187 -8.13277 -0.0122169 0.0211721 1.20606e-07
10 Si 1 -1.16457 2.01709 7.27481 -8.13277 -0.0122146 0.0211722 -4.4452e-06
11 Si 1 6.27291 6.31112 1.8187 -8.13277 -0.0122255 0.0211688 -7.31135e-06
(svn_dev) ~/K/p/b/K/S/bulk_md_lmp $ cat in.lmp | grep dump
dump my all custom 1 dump.traj id element type x y z c_peratom fx fy fz
dump_modify my element Si O
undump my
(svn_dev) ~/K/p/b/K/S/bulk_md_lmp $ cat sevenn_data/graph.yaml | grep species -A 3
species:
- Si
- O
```

If you did correctly, the `id` column of your dump text file should contain element. With this dump file, I successfully parsed `species` using `sevenn_graph_build`.

For energies, what your `dump` contains is **atomic energy** potential inferenced, not **total energy** of the system. SevenNet is designed to train on first-priciple calculations and expect to have **single total energy for each structure**. Of course, it is technically possible to compute `total_energy` by summation. However, it is not what SevenNet designed for, therefore, such routine is not implemented and you have to preprocess your dump. For example,

```python
import ase.io

atoms_list = ase.io.read(path_to_my_dump, index=':')   # read all snapshots in dump file
for atoms in atoms_list:
    atoms.info.update({'my_energy': sum([float(atomic_energy) for atomic_energy in atoms.arrays['c_energy']]))

ase.io.write('preprocessed.extxyz', atoms_list)
```

Then,
```bash
sevenn_graph_build preprocessed.extxyz 5.0 --kwargs energy_key='my_energy'
```

Above python code is I wrote directly on the comment. Modify it if it raises an error. For `stress`, you can do same approach, but if you don't want, its fine. Just make sure to set `is_train_stress=False` in your `input.yaml`, when training.
